src/DDTrace/Integrations/OpenAI/OpenAIIntegration.php

```diff
@@ -0,0 +1,1197 @@
+<?php
+
+namespace DDTrace\Integrations\OpenAI;
+
+use DDTrace\Integrations\Integration;
+use DDTrace\Log\DatadogLogger;
+use DDTrace\SpanData;
+use DDTrace\Tag;
+use DDTrace\Type;
+use DDTrace\Util\ObjectKVStore;
+use function DDTrace\dogstatsd_count;
+use function DDTrace\dogstatsd_distribution;
+use function DDTrace\dogstatsd_gauge;
+
+class OpenAIIntegration extends Integration
+{
+
+    const NAME = 'openai';
+
+    const COMMON_PARAMETERS = [
+        'best_of' => true,
+        'echo' => true,
+        'logprobs' => true,
+        'max_tokens' => true,
+        'model' => true,
+        'n' => true,
+        'presence_penalty' => true,
+        'frequency_penalty' => true,
+        'stop' => true,
+        'suffix' => true,
+        'temperature' => true,
+        'top_p' => true,
+        'user' => true,
+        'file_id' => true
+    ];
+
+    /** @var string */
+    public $serviceName;
+
+    /**
+     * @param string $apiKey The OpenAI API key to format
+     * @return string The formatted API key 'XXX...YYYY', where XXX and YYYY are respectively the first three and last
+     * four characters of the provided OpenAI API Key
+     */
+    public static function formatAPIKey(string $apiKey): string
+    {
+        return \substr($apiKey, 0, 3) . '...' . \substr($apiKey, -4);
+    }
+
+    public static function shouldSample(float $rate): bool
+    {
+        $random = \mt_rand() / \mt_getrandmax();
+        return $random <= $rate;
+    }
+
+    public function setServiceName(SpanData $span)
+    {
+        if (!empty($this->serviceName)) {
+            $span->service = $this->serviceName;
+        }
+
+        if (\dd_trace_env_config('DD_OPENAI_SERVICE')) {
+            $this->serviceName = \dd_trace_env_config('DD_OPENAI_SERVICE');
+        } elseif (\dd_trace_env_config('DD_SERVICE')) {
+            $this->serviceName = \dd_trace_env_config('DD_SERVICE');
+        } else {
+            $this->serviceName = null;
+        }
+
+        if (!empty($this->serviceName)) {
+            $span->service = $this->serviceName;
+        }
```

Comment:
Is there an else missing? Like
```suggestion
        if (!empty($this->serviceName)) {
            $span->service = $this->serviceName;
        } else {
            if (\dd_trace_env_config('DD_OPENAI_SERVICE')) {
                $this->serviceName = \dd_trace_env_config('DD_OPENAI_SERVICE');
            } elseif (\dd_trace_env_config('DD_SERVICE')) {
                $this->serviceName = \dd_trace_env_config('DD_SERVICE');
            } else {
                $this->serviceName = null;
            }

            if (!empty($this->serviceName)) {
                $span->service = $this->serviceName;
            }
        }
```

Also, I'd consider not fetching DD_SERVICE. It should respect inheritance, i.e. let's just do:
```suggestion
        if ($service = \dd_trace_env_config('DD_OPENAI_SERVICE')) {
            $span->service = $service;
        }
```

And that's it. No caching, always correct and simple.